sirepo/template/rshellweg_dump_reader.py

```diff
@@ -0,0 +1,357 @@
+    _fields_ = [("NPoints", ctypes.c_int), ("NParticles", ctypes.c_int)]
+
+
+class TField(ctypes.Structure):
+    _fields_ = [("r", ctypes.c_double), ("th", ctypes.c_double), ("z", ctypes.c_double)]
+
+
+class TStructure(ctypes.Structure):
+    _fields_ = [
+        ("ksi", ctypes.c_double),
+        ("lmb", ctypes.c_double),
+        ("P", ctypes.c_double),
+        ("dF", ctypes.c_double),
+        ("E", ctypes.c_double),
+        ("AF", ctypes.c_double),
+        ("Rp", ctypes.c_double),
+        ("B", ctypes.c_double),
+        ("alpha", ctypes.c_double),
+        ("betta", ctypes.c_double),
+        ("Ra", ctypes.c_double),
+        ("Hext", TField),
+        # Bmap replaces Hext in future versions
+        # ('Bmap', TFieldMap2D),
+        ("jump", ctypes.c_bool),
+        ("drift", ctypes.c_bool),
+        ("CellNumber", ctypes.c_int),
+    ]
+
+
+class TBeamHeader(ctypes.Structure):
+    _fields_ = [
+        ("beam_lmb", ctypes.c_double),
+        ("beam_h", ctypes.c_double),
+        ("beam_current", ctypes.c_double),
+        ("input_current", ctypes.c_double),
+    ]
+
+
+class TParticle(ctypes.Structure):
+    _fields_ = [
+        ("r", ctypes.c_double),
+        ("Th", ctypes.c_double),
+        ("beta", TField),
+        ("phi", ctypes.c_double),
+        ("z", ctypes.c_double),
+        ("beta0", ctypes.c_double),
+        ("lost", ctypes.c_int),
+    ]
+
+
+def beam_header(filename):
+    with open(filename, "rb") as f:
+        header = THeader()
+        assert f.readinto(header) == ctypes.sizeof(header)
+        return header
+
+
+def beam_info(filename, idx):
+    info = {}
+    with open(filename, "rb") as f:
+        header = THeader()
+        assert f.readinto(header) == ctypes.sizeof(header)
+        info["Header"] = header
+        structure = TStructure()
+        size = ctypes.sizeof(structure)
+        if idx > 0:
+            f.seek(idx * size, 1)
+        assert f.readinto(structure) == size
+        info["Structure"] = structure
+        if idx < header.NPoints - 1:
+            f.seek((header.NPoints - idx - 1) * size, 1)
+        beam_header = TBeamHeader()
+        size = (
+            ctypes.sizeof(beam_header) + ctypes.sizeof(TParticle()) * header.NParticles
+        )
+        if idx > 0:
+            f.seek(idx * size, 1)
+        assert f.readinto(beam_header) == ctypes.sizeof(beam_header)
+        info["BeamHeader"] = beam_header
+        particles = []
+        for _ in range(header.NParticles):
+            p = TParticle()
+            assert f.readinto(p) == ctypes.sizeof(p)
+            particles.append(p)
+        info["Particles"] = particles
+        # ensure the expected bytes are present
+        if idx < header.NPoints - 1:
+            f.seek(
+                (header.NPoints - idx - 1)
+                * (ctypes.sizeof(beam_header) + ctypes.sizeof(p) * header.NParticles),
+                1,
+            )
+        # TODO(e-carlin): is this safe?
+        # assert f.readinto(header) == 0
```

Comment:
@moellep can this be removed? It was failing because readinto read some bytes. I assume this was a catch to test whether or not we had read the full file. My guess is the input file has some additional data but the rest of the structure seems to be the same (at least the tests pass w/o any other changes...)